[StandaloneLinkContents] Fixing how color prop is handled

ashleemboyer opened the pull request:
## 🔗 Relevant links

<!--
Include links to the branch preview, Asana task, and Figma designs wherever possible to make reviewing your PR easier.
When including the preview link, make sure to remove any '.' characters from the branch name:
  - Example: ks.my-branch -> ksmy-branch
-->

- [Preview link](https://dev-portal-git-ambfix-standalone-link-contents-hashicorp.vercel.app/) 🔎

## 🗒️ What

<!--
Briefly list out the changes proposed in this PR.
-->

- Updates `StandaloneLinkContents`'s `.contents` class to inherit it's `color` when nested inside of `StandaloneLink`
- Adds `DEFAULT_COLOR_VARIANT` since it's needed in 2 places in the file now
- Simplifies the classname associated with `StandaloneLink`'s `color` prop (easier to maintain 2 references to)
- Fixes the color classes to only apply a color change on active/hover if the element it's applied to has an `href` property

## 🧪 Testing

<!--
Create a checklist for going through how to test your proposed changes. If there is anything to configure before interacting with the project in a browser, such as toggling feature flags, changing machine settings, or simulating behavior in browser dev tools, list those steps first.

- [ ] Step 1
- [ ] Step 2
- [ ] Step 3
- [ ] ...
-->

- [ ] Go to [the `StandaloneLink` Swingset page](https://dev-portal-git-ambfix-standalone-link-contents-hashicorp.vercel.app/swingset/components/standalonelink)
- [ ] The `StandaloneLink` examples should behave as expected when hovered
- [ ] The `StandaloneLink` examples should behave as expected when active
- [ ] The `StandaloneLinkContents` examples should behave as expected when hovered
- [ ] The `StandaloneLinkContents` examples should behave as expected when active

ashleemboyer commented:
🟡 We'll probably have to wait to merge this until https://github.com/hashicorp/dev-portal/pull/1830 is reviewed and merged?

zchsh commented on src/components/standalone-link/standalone-link.module.css:
```diff
@@ -32,6 +29,14 @@
+  from setting its own color.
+  */
+	& .contents {
+		color: inherit !important;
```
🚧 If possible, I think it'd be preferable to avoid the `!important` here... easier said than done though 😅 

One way to do this might be pass in an optional "inheritColor" boolean prop to `StandaloneLinkContents`. When that prop is set to `true`, the `color` classes would not be applied.

ashleemboyer replied:
I like this idea! I can't remember if I already tried it last night or not; maybe it was just something similar I tried that TypeScript yelled at me about 😅

zchsh commented on src/components/standalone-link/standalone-link.module.css:
```diff
@@ -58,22 +63,25 @@
 
-	&:hover {
+	/* only change color on active for <a href> */
+	&[href]:hover {
```
💭 For clarity here, this means the `:hover` color change effect will no longer apply to `StandaloneLinkContents`, but the underline-on-hover will still apply, right? I think we might want to retain the `:hover` for now given that's the behaviour upstream 🤔

ashleemboyer replied:
Nope! The underline-on-hover only happens in the `.root` class, which is only applied to `StandaloneLink`'s `<a>`. that's not super clear from just reading the CSS though. I'm going to rename some classes!

The purpose of this `href` specificity check is to make it so that `StandaloneLinkContents` does not change color on hover. This is because the component itself is just plain text, and not interactive. If color _does_ need to change on hover/active of an element `StandaloneLinkContents` is inside of, it can easily be layered on top of the base CSS for `StandaloneLinkContents`.

Since `StandaloneLinkContents` is not used anywhere on its own but the Swingset page, this separation shouldn't have any downstream affects.

BrandonRomano replied:
I spun up this branch here locally and was poking around here, and here's a (very sloppy) commit which illustrates my exploration here.  I believe the changes I make (which remove the inherit prop) I believe still yields a similar result.

https://github.com/hashicorp/dev-portal/commit/f3437e3180a7da744382117cbf738eb1d4dc6af1

I'll have to noodle on your approach more tomorrow.

ashleemboyer replied:
@BrandonRomano unfortunately it doesn't yield the same result, because the active/hover color style changes only apply to `.standaloneLink`. So when both `.standaloneLink` and `.standaloneLinkContents` have the same color class applied, `.standaloneLink`'s color changes on active/hover, but `.standaloneLinkContents`'s color stays the same and overwrites the color `.standaloneLink` would otherwise pass down for `.standaloneLinkContents` to inherit.

Here's how I tested:

- [ ] Go to the `StandaloneLink` Swingset page for your branch: https://dev-portal-git-brstandalone-link-diff-hashicorp.vercel.app/swingset/components/standalonelink
- [ ] Scroll down to one of the `StandaloneLink` examples
- [ ] Inspect the `a.standaloneLink` element of one of the examples
- [ ] Note what the `color` is set to

<img width="1647" alt="CleanShot 2023-04-06 at 21 32 36@2x" src="https://user-images.githubusercontent.com/43934258/230522415-7a4daaa6-c236-4d1e-8229-8fa5ee9dbc38.png">

- [ ] Inspect the `div.standaloneLinkContents` element directly inside of the `a.standaloneLink` element
- [ ] Note what the `color` is set to

<img width="1647" alt="CleanShot 2023-04-06 at 21 34 11@2x" src="https://user-images.githubusercontent.com/43934258/230522551-ad114138-9e50-47ff-bc5a-d5fbac8e479b.png">

- [ ] Inspect the `a.standaloneLink` element again
- [ ] Force the `:hover` state on the element using devtools
- [ ] Note what the `color` is now set to (it's correctly updating based on the color classname)

<img width="1645" alt="CleanShot 2023-04-06 at 21 35 23@2x" src="https://user-images.githubusercontent.com/43934258/230522703-3dc54de1-49cb-4c89-95f6-9754d01d511e.png">

- [ ] With the `:hover` state still forced on `a.standaloneLink`, inspect the `div.standaloneLinkContents` element directly inside of the `a.standaloneLink` element
- [ ] Note what the `color` is set to (it's still the base color, and not matching `a.standaloneLink`)

<img width="1645" alt="CleanShot 2023-04-06 at 21 40 10@2x" src="https://user-images.githubusercontent.com/43934258/230523211-0a655bc0-f796-47be-ae8c-6f0439a37445.png">

BrandonRomano replied:
Thank you for walking me through this here!

zchsh commented:
> 🟡 We'll probably have to wait to merge this until https://github.com/hashicorp/dev-portal/pull/1830 is reviewed and merged?

@ashleemboyer I think it might be other way around - I think we want to merge this PR first, so that the override removal in #1830 doesn't affect the underline color.

ashleemboyer commented:
> > 🟡 We'll probably have to wait to merge this until #1830 is reviewed and merged?
> 
> @ashleemboyer I think it might be other way around - I think we want to merge this PR first, so that the override removal in #1830 doesn't affect the underline color.

@zchsh ohhhhh I understand what you were calling out over there now! Yes, agreed now! We can merge this one first, then that one. 👍🏻

---

I believe the changes I've added in fd4c4b069936c1978761f1a0e8b42af2f4033aa7 resolve both of your comments now!

- Added `inheritColor`, with a fancy `developmentToast` in the case that both `color` and `inheritColor` are passed (an ambiguous case that may be hard to debug without that information)
- Removed the `!important` part, and conditionally apply the color helper class in `StandaloneLinkContents` based on `inheritColor`
- Renamed `.root` to `.standaloneLink`, and `.contents` to `.standaloneLinkContents`
- Updated the `&[href]:active`/`&[href]:hover` selectors to be clearer with `&.standaloneLink:active`/`&.standaloneLink:hover` (I didn't think this would work yesterday, but it did! 🎉)

What do you think?

BrandonRomano commented on src/components/standalone-link/index.tsx:
```diff
@@ -68,9 +80,9 @@ const StandaloneLink = ({
-				color={color}
 				icon={icon}
 				iconPosition={iconPosition}
+				inheritColor
```
It is taking me a little while to understand this `inheritColor` behavior...

It appears that on the `StandaloneLinkContents`, setting `inheritColor` simply just disallows the usage of the `color` attribute.  Are there any other behaviors here that I am missing?

It also appears that the only place that we would ever use `inheritColor=true` is on this line here (to inherit the color as set from the `StandaloneLink`).

If those is correct, could we simply just remove this `inheritColor` prop and simply just _not_ pass a color to the `StandaloneLinkContents`?

ashleemboyer replied:
@BrandonRomano this section of code already doesn't pass `color` down to `StandaloneLinkContents`. Because `StandaloneLinkContents` has a `color` prop, and by default applies the helper classes based on that prop, `StandaloneLink` needs a way to disable that behavior so it can handle the color setting _for_ `StandaloneLinkContents`.

Here's an example of how `StandaloneLinkContents` will be used on the home page:
https://github.com/hashicorp/dev-portal/pull/1823/files#diff-4cefbd1b2095426710643ee7c5e1f98fa0ebda085a4ce03b49e76b175b5f0f71R67-R72

I basically need all of `StandaloneLink`'s behavior and styling, aside from a link being rendered and color changing on active/hover. So I need `color`, `icon`, `iconPosition`, and `text` handling here, and in the other featured cards in this section of the page those props and `size` will be useful.

zchsh replied:
@BrandonRomano @ashleemboyer If I'm following Brandon's suggestion, I think an alternative to the `inheritColor` prop could be to _remove_ the `DEFAULT_COLOR_VARIANT` fallback from `StandaloneLinkContents`. Then we could not pass `color` when rendering `StandaloneLinkContents` in `StandaloneLink`. l think that's a pretty good idea, and could remove some props complexity 👍

I think the tradeoff would be it'd then be maybe easier to "misuse" `StandaloneLinkContents` - it'd be possible to render that component without a `color` prop, resulting in unexpected but difficult-to-warn-against appearance.

Personally happy with either solution 🌻
